commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexFunctions.java

```diff
@@ -354,6 +665,1057 @@ private static <R> R log(DoubleUnaryOperator log,
+                        } else {
+                            re = y / x;
+                            im = LN_2 + Math.log(x);
+                        }
+                    }
+                } else if (y <= SAFE_MIN) {
+                    // Hull et al: Assume x == 1.
+                    // True if:
+                    // E^2 > 8*sqrt(u)
+                    //
+                    // E = Machine epsilon: (1 + epsilon) = 1
+                    // u = Double.MIN_NORMAL
+                    re = Math.sqrt(y);
+                    im = Math.sqrt(y);
+                } else if (EPSILON * y - 1 >= x) {
+                    re = PI_OVER_2;
+                    im = LN_2 + Math.log(y);
+                } else if (x > 1) {
+                    re = Math.atan(y / x);
+                    final double xoy = x / y;
+                    im = LN_2 + Math.log(y) + 0.5 * Math.log1p(xoy * xoy);
+                } else {
+                    re = PI_OVER_2;
+                    final double a = Math.sqrt(1 + y * y);
+                    im = 0.5 * Math.log1p(2 * y * (y + a));
+                }
+            }
+        }
+        return action.apply(negative(real) ? Math.PI - re : re,
+                            negative(imaginary) ? im : -im);
+    }
+
+    /**
+     * Returns the
+     * <a href="http://mathworld.wolfram.com/InverseTangent.html">
+     * inverse tangent</a> of the complex number.
+     *
+     * <p>\[ \tan^{-1}(z) = \frac{i}{2} \ln \left( \frac{i + z}{i - z} \right) \]
+     *
+     * <p>The inverse hyperbolic tangent of \( z \) is unbounded along the imaginary axis and
+     * in the range \( [-\pi/2, \pi/2] \) along the real axis.
+     *
+     * <p>The inverse tangent is a multivalued function and requires a branch cut in
+     * the complex plane; the cut is conventionally placed at the line segments
+     * \( (i \infty,-i] \) and \( [i,i \infty) \) of the imaginary axis.
+     *
+     * <p>As per the C99 standard this function is computed using the trigonomic identity:
+     * \[ \tan^{-1}(z) = -i \tanh^{-1}(iz) \]
+     *
+     * @param real Real part \( a \) of the complex number \( (a +ib \).
+     * @param imaginary Imaginary part \( b \) of the complex number \( (a +ib) \).
+     * @param action Consumer for the inverse tangent of the complex number.
+     * @param <R> the return type of the supplied action.
+     * @return the object returned by the supplied action.
+     * @see <a href="http://functions.wolfram.com/ElementaryFunctions/ArcTan/">ArcTan</a>
+     */
+    public static <R> R atan(double real, double imaginary, ComplexSink<R> action) {
+        // Define in terms of atanh
+        // atan(z) = -i atanh(iz)
+        // Multiply this number by I, compute atanh, then multiply by back
+        return atanh(-imaginary, real, (r, i) -> multiplyNegativeI(r, i, action));
+    }
+
+    /**
+     * Returns the hyperbolic sine of the complex number using its real
```

Comment:
Wrong javadoc. The origin of this javadoc was from a static helper method in Complex. It may be redundant now that all functions accept a ComplexSink for the output. Previously this type of pattern was only used internally in Complex to allow computation using identities.